### PR TITLE
[Merged by Bors] - Smartstream map

### LINKED ISSUE #1172: Implement transform (map) SmartStream
Following the initial "filtering" smartstream `#[smartstream(filter)]`, we should now implement a "mapping" smartstream that takes Records as input and allows users to edit them and transform them.

Here is the current proposal for the SmartStream Map API:

```rust
use fluvio_smartstream::{smartstream, Record, RecordData};

#[smartstream(map)]
fn map(record: &Record) -> (Option<RecordData>, RecordData) {
    let key = record.key.cloned();
    let mut value = Vec::from(record.value.as_ref());

    value.make_ascii_uppercase();
    (key, value.into())
}
```

In this API:

- A Record is passed by reference, allowing full visibility into the Record and its metadata (e.g. offset)
- The input record is immutable, meaning that it is transparent to view but closed for modification
- The SmartStream returns a tuple of an optional key and a new record value

### PR DESCRIPTION
**Pull request by nicholastmosher**
Closes #1172 

This is updated since our first iteration of SmartStreams with filter. Users can write a mapping SmartStream like the following now:

```rust
use fluvio_smartstream::{smartstream, Record, RecordData};

#[smartstream(map)]
fn map(record: &Record) -> (Option<RecordData>, RecordData) {
    let key = record.key.cloned();
    let mut value = Vec::from(record.value.as_ref());

    value.make_ascii_uppercase();
    (key, value.into())
}
```

This gives the following results:

```
flvd consume test -B
one
two
Three
fouR
```

```
❯ flvd consume test -B --smart-stream=examples/target/wasm32-unknown-unknown/release/fluvio_wasm_map.wasm
ONE
TWO
THREE
FOUR
```

### CODE REVIEW

**Comment by nicholastmosher** [comment]
I am thinking of writing another example for this, possibly something to do with JSON, similar to the filter_json example

**Comment by nicholastmosher** [comment]
@sehz this has been updated with the latest API we discussed in #1172, it is now ready to review

**Review by sehz — CHANGES_REQUESTED** [review]
LGTM.  Nice job. To finish this, 
- bump up version
- add to changelog

File: src/dataplane-protocol/src/record.rs [hunk 1]
```diff
@@ -82,7 +82,7 @@ pub type DefaultAsyncBuffer = RecordData;
 /// [the Producer API] as an example.
 ///
 /// [the Producer API]: https://docs.rs/fluvio/producer/TopicProducer::send
-#[derive(Default)]
+#[derive(Clone, Default)]
```
Comment (sehz) [inline src/dataplane-protocol/src/record.rs:85]: this requires bump up version for dataplane-protocol

**Review by sehz — CHANGES_REQUESTED** [review]

File: src/smartstream/Cargo.toml [hunk 2]
```diff
@@ -18,4 +18,4 @@ crate-type = ['lib']
 
 [dependencies]
 fluvio-dataplane-protocol = { version = "0.5.1", path = "../dataplane-protocol", default-features = false }
```
Comment (sehz) [inline src/smartstream/Cargo.toml:20]: this should be 0.5.2?

Comment (nicholastmosher) [reply src/smartstream/Cargo.toml]: Ah yep. Updated dependencies

File: src/auth/Cargo.toml [hunk 3]
```diff
@@ -22,7 +22,7 @@ tracing-futures = "0.2.4"
 x509-parser = "0.9.1"
 
 fluvio-controlplane-metadata = { version = "0.9.1", path = "../controlplane-metadata" }
-dataplane = { version = "0.5.1", path = "../dataplane-protocol", package = "fluvio-dataplane-protocol" }
+dataplane = { version = "0.5.2", path = "../dataplane-protocol", package = "fluvio-dataplane-protocol" }
```
Comment (sehz) [inline src/auth/Cargo.toml:25]: no need for updating this

File: src/client/Cargo.toml [hunk 4]
```diff
@@ -44,7 +44,7 @@ fluvio-types = { version = "0.2.1", features = ["events"], path = "../types" }
 fluvio-sc-schema = { version = "0.8.1", path = "../sc-schema", default-features = false }
 fluvio-socket = { path = "../socket", version = "0.8.1" }
 fluvio-protocol = { path = "../protocol", version = "0.5.1" }
-dataplane = { version = "0.5.1", path = "../dataplane-protocol", package = "fluvio-dataplane-protocol" }
+dataplane = { version = "0.5.2", path = "../dataplane-protocol", package = "fluvio-dataplane-protocol" }
```
Comment (sehz) [inline src/client/Cargo.toml:47]: no need

File: src/controlplane-metadata/Cargo.toml [hunk 5]
```diff
@@ -27,7 +27,7 @@ flv-util = { version = "0.5.0" }
 fluvio-types = { version = "0.2.0", path = "../types" }
 fluvio-stream-model = { path = "../stream-model", version = "0.5.0" }
 fluvio-protocol = { path = "../protocol", version = "0.5.1" }
-dataplane = { version = "0.5.1", path = "../dataplane-protocol", package = "fluvio-dataplane-protocol" }
+dataplane = { version = "0.5.2", path = "../dataplane-protocol", package = "fluvio-dataplane-protocol" }
```
Comment (sehz) [inline src/controlplane-metadata/Cargo.toml:30]: no need

File: src/controlplane/Cargo.toml [hunk 6]
```diff
@@ -19,4 +19,4 @@ tracing = "0.1.19"
 fluvio-types = { path = "../types", version = "0.2.0" }
 fluvio-controlplane-metadata = { path = "../controlplane-metadata", version = "0.9.1" }
 fluvio-protocol = { path = "../protocol", version = "0.5.1" }
-dataplane = { version = "0.5.1", path = "../dataplane-protocol", package = "fluvio-dataplane-protocol" }
```
Comment (sehz) [inline src/controlplane/Cargo.toml:22]: no need

File: src/sc-schema/Cargo.toml [hunk 7]
```diff
@@ -25,4 +25,4 @@ static_assertions = "1.1.0"
 fluvio-types = { version = "0.2.0", path = "../types" }
 fluvio-controlplane-metadata = { version = "0.9.1", default-features = false, path = "../controlplane-metadata" }
 fluvio-protocol = { path = "../protocol", version = "0.5.1" }
-dataplane = { version = "0.5.1", path = "../dataplane-protocol", package = "fluvio-dataplane-protocol" }
```
Comment (sehz) [inline src/sc-schema/Cargo.toml:28]: no need

File: src/sc/Cargo.toml [hunk 8]
```diff
@@ -55,7 +55,7 @@ k8-metadata-client = { version = "3.0.0" }
 k8-types = { version = "0.2.1", features = ["app"] }
 fluvio-protocol = { path = "../protocol", version = "0.5.1" }
 fluvio-socket = { path = "../socket", version = "0.8.1" }
-dataplane = { version = "0.5.1", path = "../dataplane-protocol", package = "fluvio-dataplane-protocol" }
+dataplane = { version = "0.5.2", path = "../dataplane-protocol", package = "fluvio-dataplane-protocol" }
```
Comment (sehz) [inline src/sc/Cargo.toml:58]: no need

**Review by sehz — CHANGES_REQUESTED** [review]
only need to update smartstream's dependencies

File: src/spu-schema/Cargo.toml [hunk 9]
```diff
@@ -23,4 +23,4 @@ static_assertions = "1.1.0"
 
 # Fluvio dependencies
 fluvio-protocol = { path = "../protocol", version = "0.5.1" }
-dataplane = { version = "0.5.1", path = "../dataplane-protocol", package = "fluvio-dataplane-protocol" }
+dataplane = { version = "0.5.2", path = "../dataplane-protocol", package = "fluvio-dataplane-protocol" }
```
Comment (sehz) [inline src/spu-schema/Cargo.toml:26]: this doesn't need to be updated

File: src/spu/Cargo.toml [hunk 10]
```diff
@@ -50,7 +50,7 @@ fluvio-controlplane-metadata = { version = "0.9.1", path = "../controlplane-meta
 fluvio-spu-schema = { version = "0.6.0", path = "../spu-schema", features = ["file"]}
 fluvio-protocol = { path = "../protocol", version = "0.5.1" }
 fluvio-socket = { path = "../socket", version = "0.8.1", features = ["file"] }
-dataplane = { version = "0.5.1", path = "../dataplane-protocol", package = "fluvio-dataplane-protocol" , features=["file"]}
+dataplane = { version = "0.5.2", path = "../dataplane-protocol", package = "fluvio-dataplane-protocol" , features=["file"]}
```
Comment (sehz) [inline src/spu/Cargo.toml:53]: same here

File: src/storage/Cargo.toml [hunk 11]
```diff
@@ -35,7 +35,7 @@ derive_builder = "0.10.2"
 fluvio-types = { version = "0.2.0", path = "../types" }
 fluvio-future = { version = "0.3.2", features = ["fs", "mmap","zero_copy"] }
 fluvio-protocol = { path = "../protocol", version = "0.5.1" }
-dataplane = { version = "0.5.1", path = "../dataplane-protocol", package = "fluvio-dataplane-protocol", features = ["file"] }
+dataplane = { version = "0.5.2", path = "../dataplane-protocol", package = "fluvio-dataplane-protocol", features = ["file"] }
```
Comment (sehz) [inline src/storage/Cargo.toml:38]: same

**Review by sehz — CHANGES_REQUESTED** [review]

**Review by sehz — APPROVED** [review]
LGTM

**Pull request closed without merging**